common-controller/internal/operator/controllers/dp/gatewayclass_controller.go

```diff
@@ -0,0 +1,147 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	k8client "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gwapiv1b1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+	constants "github.com/wso2/apk/common-controller/internal/operator/constant"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	gatewayRateLimitPolicyIndex = "gatewayRateLimitPolicyIndex"
+	gatewayAPIPolicyIndex       = "gatewayAPIPolicyIndex"
+)
+
+var (
+	setReadiness sync.Once
+)
+
+// GatewayClassReconciler reconciles a Gateway object
+type GatewayClassReconciler struct {
+	client        k8client.Client
+	statusUpdater *status.UpdateHandler
+	mgr           manager.Manager
+}
+
+// NewGatewayClassController creates a new GatewayClass controller instance. GatewayClass Controllers watches for gwapiv1b1.GatewayClass.
+func NewGatewayClassController(mgr manager.Manager, statusUpdater *status.UpdateHandler) error {
+	r := &GatewayClassReconciler{
+		client:        mgr.GetClient(),
+		statusUpdater: statusUpdater,
+		mgr:           mgr,
+	}
+
+	c, err := controller.New(constants.GatewayClassController, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		loggers.LoggerAPKOperator.ErrorC(logging.PrintError(logging.Error2663, logging.BLOCKER,
+			"Error creating GatewayClass controller: %v", err.Error()))
+		return err
+	}
+
+	if err := c.Watch(source.Kind(mgr.GetCache(), &gwapiv1b1.GatewayClass{}), &handler.EnqueueRequestForObject{}); err != nil {
+		loggers.LoggerAPKOperator.ErrorC(logging.PrintError(logging.Error2639, logging.BLOCKER,
+			"Error watching GatewayClass resources: %v", err.Error()))
+		return err
+	}
+
+	loggers.LoggerAPKOperator.Info("GatwayClasses Controller successfully started. Watching GatewayClass Objects...")
+	return nil
+}
+
+//+kubebuilder:rbac:groups=dp.wso2.com,resources=gatewayclasses,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=dp.wso2.com,resources=gatewayclasses/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=dp.wso2.com,resources=gatewayclasses/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (gatewayClassReconciler *GatewayClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// Check whether the Gateway CR exist, if not consider as a DELETE event.
+	loggers.LoggerAPKOperator.Info("Reconciling gateway class...")
+	var gatewayClassDef gwapiv1b1.GatewayClass
+	gatewayClassReconciler.client.Get(ctx, req.NamespacedName, &gatewayClassDef);
```

Comment:
there could be instances where this call will fail, right? Like reconcillation triggered for a deletion. Don't we need to handle it?

Reply:
Handled. If not found error then return the method. If the error is some other unexpected we will requeue after one second